# Wind prediction - Second assignment

## Authors

David Moreno Maldonado 100441714     
Inés Fernández Campos 100443936

## 0. Preliminaries

In [2]:
# Import some libraries
import os
import numpy as np              
import pandas as pd
import matplotlib.pyplot as plt 

os.getcwd()

'/home/fddcampos/Documents/uc3m/2_term/BDINTELLIGENCE/practicas_big_data_intelligence/assignment_2'

The "wind_pickle" file contains data in a binary format called "Pickle". Pickle data loads faster than text data.

In [3]:
data = pd.read_pickle('wind_pickle.pickle')

You can visualize the attributes in the dataset. Very important, the output attribute (i.e. the value to be predicted, **energy**, is the first attribute). **Steps** represents the hours in advance of the forecast. We will not use this variable here.

In [12]:
# The dataset contains 5937 instances and 556 attributes (including the outcome to be predicted)
print(data.shape)
# data.columns.values.tolist() 

(5937, 556)


We now remove the columns that cannot be used for training the models from the DataFrame

In [ ]:
# Steps, month, day, hour, year should be removed, they cannot be used for training the models
to_remove = ['steps', 'month', 'year', 'day', 'hour']
for m in to_remove: df = df.drop(m, 1)
    


In [7]:
from numpy.random import randint

# we add na values at random
my_NIA = 100443936
np.random.seed(my_NIA)

how_many_nas = round(data.shape[0]*data.shape[1]*0.05)
print('Lets put '+str(how_many_nas)+' missing values \n')
x_locations = randint(0, data.shape[0], size=how_many_nas)
y_locations = randint(6, data.shape[1]-1, size=how_many_nas)

for i in range(len(x_locations)):
    data.iat[x_locations[i], y_locations[i]] = np.nan

Lets put 165049 missing values 



From this point on, the file wind_pickle_with_nan should be used.

In [4]:
data = pd.read_pickle('wind_pickle_with_nan.pickle')

## Data

In [11]:
data[data.columns.difference(['energy'])].shape
data.shape

(5937, 551)

We are going to use train/test for model evaluation (outer) and train/validation for hyperparameter tuning (inner), as follows:     
1. Train partition: the first two years of data. Given that there are 6 years worth of data, we will use the first 2/6 of the instances for training.     
2. Validation partition: the second two years of data. 
3. Test partition: the remaining data    


In [40]:
from sklearn import model_selection
import random

random.seed(1234)

x = data[data.columns.difference(['energy'])]
y = data['energy']

#Split in subsets using holdout 2/6 for training, 2/6 validation, 2/6 for testing
x_train, x_test, y_train, y_test = model_selection.train_test_split(x, y, train_size=(2/3), random_state=0)

# -1 for training, 0 for validation 
part = [-1, -1, 0, 0]
random.shuffle(part)
tr_val_partition = model_selection.PredefinedSplit(part)

print('whole dataset: ', x.shape, y.shape)
print('train partition: ', x_train.shape, y_train.shape)
print('test partition: ', x_test.shape, y_test.shape)

whole dataset:  (5937, 550) (5937,)
train partition:  (3958, 550) (3958,)
test partition:  (1979, 550) (1979,)


Now we convert the training and test sets from Pandas DataFrames to Numpy matrices, so that they can be used by scikit-learn.

In [24]:
mat_x_train = np.matrix(x_train)
mat_y_train = np.matrix(y_train)
mat_x_test = np.matrix(x_test)
mat_y_test = np.matrix(y_test)

# 1. MODEL SELECTION AND HYPER-PARAMETER TUNING

Train a KNN model with default hiper-parameters

In [ ]:
#<WRITE CODE HERE FOR KNN>

Do hyper-parameter tuning for KNN. Can you improve results? Note: I will specially value if you use model based optimization (**skopt**) or **Optuna**.

In [ ]:
#<WRITE CODE HERE FOR KNN>

Now, train ensemble techniques (Random Forest and Gradient Boosting) without and with hyper-parameter tuning, and compare them using the test partition.

In [ ]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.ensemble import GradientBoostingRegressor

rf = RandomForestRegressor()
# help('sklearn.ensemble.RandomForestRegressor')

gb = GradientBoostingRegressor()
# help('sklearn.ensemble.GradientBoostingRegressor')

# YOU CAN USE ADVANCED IMPLEMTATIONS OF GRADIENT BOOSTING: XGBOOST, LIGHTBOOST, CATBOOST, ...

In [ ]:
#<USE AS MANY CELLS AS YOU NEED>

# 2. ATTRIBUTE SELECTION

You have to answer the following questions: 

- Are all 550 input attributes actually necessary in order to get a good model? Is it possible to have an accurate model that uses fewer than 550 variables? How many? 
- Is it enough to use only the attributes for the actual Sotavento location? (13th location in the grid)

In [ ]:
#<USE AS MANY CELLS AS YOU NEED>